### PR TITLE
Update AppDataDirGuesser for BaseDexClassLoader

### PR DESCRIPTION
**Pull request by agampe**
Since Jelly Bean 4.3 the toString of a BaseDexClassLoader has had a
different format. Update the AppDataDirGuesser to understand that
format.

### CODE REVIEW

File: dexmaker/src/main/java/com/google/dexmaker/AppDataDirGuesser.java [hunk 1]
```diff
@@ -61,12 +61,55 @@ private String getPathFromThisClassLoader(ClassLoader classLoader,
...
 
         // Parsing toString() method: yuck.  But no other way to get the path.
-        // Strip out the bit between square brackets, that's our path.
         String result = classLoader.toString();
-        int index = result.lastIndexOf('[');
-        result = (index == -1) ? result : result.substring(index + 1);
-        index = result.indexOf(']');
-        return (index == -1) ? result : result.substring(0, index);
+
+        // There are differences in Android versions. For old versions, this is just about taking
```
Comment (swankjesse) [inline dexmaker/src/main/java/com/google/dexmaker/AppDataDirGuesser.java]: Could you do a few things to make this easier to maintain?

1. Pull it out into its own method
2. Add a test case for the old form & new form?
3. Add an example of the toString output here?

**Comment by swankjesse** [comment]
Thanks for the fix! In response to this I'm asking for even more code to make it easier to keep this working going forward.

**Comment by agampe** [comment]
I didn't add tests as none of this was exposed before. Only guessPath was tested, and that doesn't change, so I kept it intentionally small (but more intention-revealing than a patch suggested on the old code.google.com website). I'll be happy to refactor.

**Comment by agampe** [comment]
Please take a look at the refactor. I've added some tests that hopefully will point out when future changes break this.

The AppDataDirGuesser tests succeeded on an N10 with Android 4.2.2, 4.3, 4.4.4, 5.0, 5.1.

**Comment by swankjesse** [comment]
@apyousef @dshirley can you guys review this?

**Comment by dshirley** [comment]
The code looks like it's accomplishing its goal, but in general the guessing code is pretty hairy.

Taking a step back, what is the reason DexMaker needs to guess a data directory path in the first place?  Can we force the user to specify a dexCache directory or the dexmaker.dexcache property?  I assume there's some backward compatibility issues we'd need to worry about.

@agampe can you fill in some details?  

I'm not going to block this PR, but it would be good to have some color on why this feature was added and if it can be simplified in the future.

Thanks!

**Comment by agampe** [comment]
I got complaints from coworkers about unexpected "dexCache == null" exceptions. The reason is that developers usually do *not* read the documentation very well, and the dex cache is somewhat of an implementation detail. People focus on Mockito as the "frontend" to DexMaker. One could argue it's a Mockito bug of hiding (and not setting) the configuration.

The AppDataDirGuesser is pretty old, and has the function of being a fallback. As such it seemed reasonably to bring it to the present and understand the different formats. Note that there has been a proposed fix on the code.google.com page of dexmaker that's pretty small, in essence a three-liner that uses string replacement operations to accomplish the same. I found it better to be more explicit about it and actually parse this "properly."

I do *not* know what the proper approach is. I think backwards compatibility is not a major issue, as the guesser has been broken since 4.3. So anyone running tests recently should have hit this problem anyways. In that sense it sounds acceptable to just remove it completely.

**Comment by dshirley** [comment]
Thanks for the background.  I think you're right the bug is that the Mockito adapter doesn't force specifying this.  At the same time, it would be a pretty wonky API if we did require it.

Lets stick with what you have.  As you said, this has been broken for a while and nobody noticed. Lets not over think it.

Thanks @swankjesse for taking the charge on the bulk of the review.

I'm merging it.
